File: 1653-number-of-good-leaf-nodes-pairs/1653-number-of-good-leaf-nodes-pairs.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.ans = 0

    def dfs(self, root, distance):
        if not root:
            return []
        if not root.left and not root.right:
            return [1]

        left = self.dfs(root.left, distance)
        right = self.dfs(root.right, distance)

        for x in left:
            for y in right:
                if x + y <= distance:
                    self.ans += 1

        v = []
        for x in left:
            v.append(x + 1)
        for y in right:
            v.append(y + 1)

        return v

    def countPairs(self, root, distance):
        self.dfs(root, distance)
        return self.ans
        

File: 1653-number-of-good-leaf-nodes-pairs/1653-number-of-good-leaf-nodes-pairs.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int ans;
    List<Integer> dfs(TreeNode root, int d) {
        List<Integer> v = new ArrayList<>();
        if(root == null) return v;
        if(root.left == null && root.right == null) {
            v.add(1);
            return v;
        }

        List<Integer> left = dfs(root.left, d);
        List<Integer> right = dfs(root.right, d);

        for(int x : left) {
            for(int y : right) {
                if(x + y <= d) ans++;
            }
        }

        for(int x : left) v.add(x + 1);
        for(int x : right) v.add(x + 1);

        return v;
    }
    
    public int countPairs(TreeNode root, int distance) {
        ans = 0;
        dfs(root, distance);
        return ans;
    }
}